src/H5Dmpio.c

```diff
@@ -1783,134 +1789,90 @@ H5D__link_chunk_filtered_collective_io(H5D_io_info_t *io_info, H5D_dset_io_info_
-        HASH_CLEAR(hh, chunk_hash_table);
+        HASH_CLEAR(hh, chunk_list.chunk_hash_table);
 
         /* All ranks now collectively re-allocate file space for all chunks */
-        if (H5D__mpio_collective_filtered_chunk_reallocate(chunk_list, chunk_list_num_entries,
-                                                           rank_chunks_assigned_map, io_info, &index_info,
-                                                           mpi_rank, mpi_size) < 0)
+        if (H5D__mpio_collective_filtered_chunk_reallocate(&chunk_list, rank_chunks_assigned_map, io_info,
+                                                           &index_info, mpi_rank, mpi_size) < 0)
             HGOTO_ERROR(H5E_DATASET, H5E_WRITEERROR, FAIL,
                         "couldn't collectively re-allocate file space for chunks")
 
-        /* If this rank has any chunks selected, create a MPI type for collectively
-         * writing out the chunks to file. Otherwise, the rank contributes to the
-         * collective write with a none type.
-         */
-        if (H5D__mpio_collective_filtered_io_type(chunk_list, chunk_list_num_entries, io_info->op_type,
-                                                  &mem_type, &mem_type_is_derived, &file_type,
-                                                  &file_type_is_derived) < 0)
-            HGOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL,
-                        "couldn't create MPI type for writing filtered chunks")
-
-        mpi_buf_count = (file_type_is_derived || mem_type_is_derived) ? 1 : 0;
-
-        /* Setup contig storage info for I/O operation */
-        if (chunk_list_num_entries) {
-            /*
-             * Override the write buffer to point to the first
-             * chunk's data buffer
-             */
-            io_info->base_maddr.cvp = chunk_list[0].buf;
-
-            /*
-             * Setup the base storage address for this operation
-             * to be the first chunk's file address
-             */
-            io_info->store_faddr = chunk_list[0].chunk_new.offset;
-        }
-        else {
-            io_info->base_maddr.cvp = &fake_buf;
-            io_info->store_faddr    = 0;
-        }
-
-        /* Perform I/O */
-        if (H5D__final_collective_io(io_info, mpi_buf_count, file_type, mem_type) < 0)
-            HGOTO_ERROR(H5E_IO, H5E_CANTGET, FAIL, "couldn't finish MPI-IO")
+        /* Perform vector I/O on chunks */
+        if (H5D__mpio_collective_filtered_vec_io(&chunk_list, io_info->f_sh, io_info->op_type) < 0)
+            HGOTO_ERROR(H5E_DATASET, H5E_WRITEERROR, FAIL, "couldn't perform vector I/O on filtered chunks")
 
         /* Free up resources in anticipation of following collective operation */
-        for (i = 0; i < chunk_list_num_entries; i++) {
-            if (chunk_list[i].buf) {
-                H5MM_free(chunk_list[i].buf);
-                chunk_list[i].buf = NULL;
+        for (size_t i = 0; i < chunk_list.num_chunk_infos; i++) {
```

Comment:
I was under the impression that our style conventions said not to place variable declarations in a loop statement like this. At least I haven't seen that anywhere else. If everyone wants to start doing this I don't have a problem with it , as long as there's not some obscure compiler we support that doesn't support this.